File: 2018/ruby/1_chronal_calibration.rb
#!/usr/bin/env ruby
freq = 0
File.open('../1', 'r') do |file|
  file.each do |freq_change|
    freq += Integer(freq_change)
  end
end
puts freq

require 'set'
freq = 0
reached_freqs = Set[0]
File.open('../1', 'r') do |file|
  found = false
  until found
    file.seek(0)
    while !found && (freq_change = file.gets)
      freq += Integer(freq_change)
      found = true unless reached_freqs.add?(freq)
    end
  end
end
puts freq

File: 2018/elixir/1_chronal_calibration.exs
#!/usr/bin/env elixir
defmodule ChronalCalibration do
  def solve_b(freq_changes) do
    freq_changes
    |> Stream.cycle()
    |> Enum.reduce_while(
      {0, MapSet.new([0])},
      fn freq_change, {prev_freq, reached_freqs} = _acc ->
        freq = prev_freq + freq_change

        if freq in reached_freqs do
          {:halt, freq}
        else
          reached_freqs = MapSet.put(reached_freqs, freq)
          {:cont, {freq, reached_freqs}}
        end
      end
    )
  end

  def read_file(filename) do
    filename
    |> File.stream!()
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.to_integer/1)
  end
end

"../1" |> ChronalCalibration.read_file() |> Enum.sum() |> IO.puts()
"../1" |> ChronalCalibration.read_file() |> ChronalCalibration.solve_b() |> IO.puts()
